### PR TITLE
Feature/access control

### PR DESCRIPTION
**Pull request by teasmade**
- Adds role based protection to client page routes
- Adds role based protection to coreapi resolvers
- Adapts backend tests with mocked context and mock user whose role can be changed
- Adds backend tests for positive and negative cases of access-control by role (e.g. CAN get patients by name as DOCTOR, CANNOT get patients by name as AGENT)
- Adapts backend tests to use the native ExecutionResult type from Graphql, adapts negative result tests to handle the edge case that, as per the GQL spec ( https://graphql.org/learn/response/ ), a "partial" response can contain both .errors and .data properties.

### CODE REVIEW

**Review by Adri-bis — DISMISSED** [review]

**Review by julien44830 — DISMISSED** [review]

**Review by julien44830 — DISMISSED** [review]

**Review by Adri-bis — DISMISSED** [review]

File: client/src/main.tsx [hunk 1]
```diff
@@ -34,19 +34,35 @@ const router = createBrowserRouter([
...
+          <ProtectedRoute allowedRoles={[RoleCode.Agent]}>
+            <AgentHome />
+          </ProtectedRoute>
+        )
       },
       {
         path: 'dossiers',
-        element: <DossierBrowser />
+        element: (
+          <ProtectedRoute allowedRoles={[RoleCode.Agent]}>
```
Comment (Acetoshi) [inline client/src/main.tsx]: Non, Doctor.

File: core_api/src/modules/consultation/consultation.resolver.ts [hunk 2]
```diff
@@ -1,9 +1,10 @@
-import { Consultation } from '../entities.index';
-import { Resolver, Query, Arg } from 'type-graphql';
+import { Consultation, RoleCode } from '../entities.index';
+import { Resolver, Query, Arg, Authorized } from 'type-graphql';
 
 @Resolver(Consultation)
 export default class ConsultationResolver {
   // TODO : rescrtict access to role === doctor
```
Comment (Acetoshi) [inline core_api/src/modules/consultation/consultation.resolver.ts]: Remove the TODO, it's done.

File: core_api/src/modules/patient/patient.resolver.ts
Comment (Acetoshi) [inline core_api/src/modules/patient/patient.resolver.ts:1]: perfect

File: core_api/src/modules/patient/patient.test.ts [hunk 3]
```diff
@@ -12,95 +14,196 @@ export const GET_PATIENTS_BY_NAME = gql`
   }
 `;
 
+const mockUser: MockUser = {
+  id: '1',
+  role: {
+    code: null // to change for each test
```
Comment (Acetoshi) [inline core_api/src/modules/patient/patient.test.ts:20]: clever

File: core_api/src/modules/patient/patient.test.ts [hunk 4]
```diff
@@ -12,95 +14,196 @@ export const GET_PATIENTS_BY_NAME = gql`
...
+    const result: ExecutionResult = await graphql({
+      schema: schema,
+      source: print(GET_PATIENTS_BY_NAME),
+      contextValue,
+      variableValues: { search: searchValue }
+    });
+
+    expect(result.errors).toEqual(expect.any(Array));
+    expect(result.data).toBeNull();
+  });
```
Comment (Acetoshi) [inline core_api/src/modules/patient/patient.test.ts:133]: great !

File: core_api/src/modules/patient/patient.test.ts [hunk 5]
```diff
@@ -12,95 +14,196 @@ export const GET_PATIENTS_BY_NAME = gql`
...
       expect(typeof patient.id).toBe('string');
     });
   });
 
   // Edge cases
   it('returns an empty array if no patient was found', async () => {
+    mockUser.role.code = RoleCode.DOCTOR;
     const searchValue = 'noPatientMatchesThisString';
-    const result = (await graphql({
+    const result: ExecutionResult = await graphql({
```
Comment (Acetoshi) [inline core_api/src/modules/patient/patient.test.ts:174]: ok, it's more robust this way, i suppose ?

File: core_api/src/modules/role/role.test.ts [hunk 6]
```diff
@@ -12,16 +14,30 @@ const GET_ROLES = gql`
...
+  it('Can get all roles as an ADMIN', async () => {
+    mockUser.role.code = RoleCode.ADMIN;
+    const result: ExecutionResult = (await graphql({
       schema: schema,
-      source: print(GET_ROLES)
+      source: print(GET_ROLES),
+      contextValue
     })) as { data: { roles: Array<unknown> } };
-    expect(result.data.roles).toEqual(expect.any(Array));
+    expect(result.data?.roles).toEqual(expect.any(Array));
```
Comment (Acetoshi) [inline core_api/src/modules/role/role.test.ts:41]: do we want to test that other users CANNOT query roles ? maybe a bit pointless

**Review by Acetoshi — CHANGES_REQUESTED** [review]
I can't see anything breaking apart from the dossier browser (see main.tsx), great job.

File: core_api/src/modules/role/role.test.ts (hunk 6 above)
Comment (teasmade) [reply core_api/src/modules/role/role.test.ts:41]: Yes my thought was that patients, as sensitive data, were the ones to protect as fully as possible with all the combinations of CAN / CANNOT query... same will apply to tests for users and consultations.

For these parameter type entities (roles, departments etc.) wasn't sure whether adding the inverse (CANNOT) test would be worth the increase in repeated code and execution time, so left it like this as a potential discussion point.

I can see an argument for repeating every single role explicitly every time, in that this would then make tests fail when any change / addition / removal is made to any kind of resolver protection, which would then be a second layer of sense-checking for the developer implementing a new feature.

🤷‍♂️

**Review by Acetoshi — APPROVED** [review]
thanks for the upadte

**Review by Adri-bis — APPROVED** [review]

**Review by julien44830 — APPROVED** [review]
